File: Python/tasks/atisLabel.py
import sys
import os
from random import shuffle,random
sys.path.insert(0,'../models/')
sys.path.insert(0,'../util/')
sys.path.insert(0,'../methods/')

import RNN
import sgd_const_lr
import ssd_const_lr
import ssd_rms
import loader
import time
import numpy as np

'''
>>> Use RNN to train token labelling problem on dataset ATIS
>>> mode: string. Specify an optimizer like 'sgd_const_lr'
>>> U_lr, W_lr, V_lr, s_lr: float. The learning rate for different parameter sets
>>> outfile: string. Set the output file we put the result in
'''

if len(sys.argv)!=8:
	print 'Usage: python atisLabel.py <mode> <U_lr> <W_lr> <V_lr> <s_lr> <a/w> <output_file>'
	exit(0)

maketest=True
decreasingLR=True
param={}
param['mode']=sys.argv[1]
param['trainXFile']='../atis/train_word1000.csv'	#Sentences in training set
param['trainYFile']='../atis/train_label1000.csv'	#Labels in training set
param['dictFile']='../atis/dict10.csv'			#Pretrained dict to map words to vectors
param['vectorDim']=100					#The dimension of word vectors
param['window']=3					#The length of the sliding windows
param['inputs']=param['vectorDim']*param['window']	#Input dimension
param['hiddens']=50					#Dimension of hidden state
param['outputs']=128					#Output dimension
param['batch']=25					#Batch size
param['nEpoch']=1000					#Number of training epochs
param['learnRates']={'U':float(sys.argv[2]),'W':float(sys.argv[3]),
	'V':float(sys.argv[4]),'s':float(sys.argv[5])}	#Learning rate of each parameter matrices
param['leftPad']=param['window']/2			#Padding length
param['outputMode']=sys.argv[6]				#Append or rewrite the output file
param['outfile']='train_' + sys.argv[7] 	   #Output file containing the results on training set
param['test_outfile']='test_' + sys.argv[7]	       #Output file containing the results on test set
param['numberOfSave']=9					#The index of the saved file
param['alpha']=random()					#A parameter used in RMSprop optimizer
param['dcrorcst']='cst'				      #Decreasing learning rate or const learning rate
if decreasingLR:
	param['dcrorcst'] = 'dcr'

train_index=loader.loadData(param['trainXFile'])
train_label=loader.loadData(param['trainYFile'])
dictionary=loader.loadDict(param['dictFile'])
dictionary[-1]=np.random.randn(param['vectorDim'])*0.5

rnn=RNN.RNN(param['inputs'],param['hiddens'],param['outputs'])

if maketest:
	lmbda = random()
	shuffle(train_index, lambda: lmbda)
	shuffle(train_label, lambda: lmbda)

train_num=len(train_index)
trainX=[]
trainY=[]

##########################################
#                                        #
#         Preprocessing the data         #
#                                        #
##########################################

for i in xrange(train_num):
	sentence_len=len(train_index[i])
	single_input=np.zeros([sentence_len,param['inputs']])
	single_label=np.zeros([sentence_len,param['outputs']])

	for p in xrange(sentence_len+param['window']-1):
		toFill=dictionary[-1]
		if p>=param['leftPad'] and p<sentence_len+param['leftPad']:
			word=train_index[i][p-param['leftPad']]
			if dictionary.has_key(word)==False:
				dictionary[word]=np.random.randn(param['vectorDim'])*0.05
			toFill=dictionary[word]
		startline=max(0,p-param['window']+1)
		endline=min(sentence_len-1,p)
		for line in range(startline,endline+1):
			col=p-line
			single_input[line,col*param['vectorDim']:(col+1)*param['vectorDim']]=toFill

	for w in xrange(sentence_len):
		single_label[w,train_label[i][w]]=1
	trainX.append(single_input)
	trainY.append(single_label)

if maketest:
	ttX=trainX
	ttY=trainY
	trainX=ttX[0:900]
	trainY=ttY[0:900]
	testX=ttX[900:1000]
	testY=ttY[900:1000]

	train_num=900
	test_num=100
	test_results='''mode=%s,U_lr=%.4f,W_lr=%.4f,V_lr=%.4f,s_lr=%.4f\n'''%(param['mode'],param['learnRates']['U'],
		param['learnRates']['W'],param['learnRates']['V'],param['learnRates']['s'])

results='''mode=%s,U_lr=%.4f,W_lr=%.4f,V_lr=%.4f,s_lr=%.4f\n'''%(param['mode'],param['learnRates']['U'],
	param['learnRates']['W'],param['learnRates']['V'],param['learnRates']['s'])
print results,

######################
#                    #
#     Start SGD      #
#                    #
######################

if param['mode']=='sgd_const_lr':
	begin=time.time()
	lr_decay=1
	for epoch in xrange(param['nEpoch']):
		if decreasingLR:
			lr_decay=epoch+1
		if maketest:
			test_err_total=0.0
			for index in xrange(test_num):
				if len(train_index[index+900])>0:
					test_err=sgd_const_lr.sgd(rnn,testX[index],testY[index],False)
					test_err_total+=test_err
			test_err_total*=10
			test_results+='|%f'%(test_err_total)
			os.system('''echo '|%s' >> results_saved_%s_test_%d_%s_%.4f_%.4f_%.4f_%.4f.log'''%(test_err_total,param['dcrorcst'],param['numberOfSave'],param['mode'],param['learnRates']['U'],param['learnRates']['W'],param['learnRates']['V'],param['learnRates']['s']))
		err_total=0.0
		for index in xrange(train_num):
			if len(train_index[index])>0:
				err=sgd_const_lr.sgd(rnn,trainX[index],trainY[index])
				err_total+=err
			if index%param['batch']==0:
				rnn.update(param['learnRates'],lr_decay)
		rnn.update(param['learnRates'],lr_decay)
		results+='|%f'%err_total
		os.system('''echo '|%s' >> results_saved_%s_train_%d_%s_%.4f_%.4f_%.4f_%.4f.log'''%(err_total,param['dcrorcst'],param['numberOfSave'],param['mode'],param['learnRates']['U'],param['learnRates']['W'],param['learnRates']['V'],param['learnRates']['s']))
		print 'epoch %d completed!'%epoch
		if maketest:
			print 'total train error=%f'%err_total
			print 'total test error=%f'%test_err_total
		else:
			print 'total error=%f'%err_total
	results+='\n'
	end=time.time()
	results+='time=%.2f\n'%(end-begin)

	if maketest:
		test_results+='\n'
		test_results+='time=%.2f\n'%(end-begin)

######################
#                    #
#     Start SSD      #
#                    #
######################

if param['mode']=='ssd_const_lr':
	begin=time.time()
	lr_decay=1
	for epoch in xrange(param['nEpoch']):
		if decreasingLR:
			lr_decay=epoch+1
		if maketest:
			test_err_total=0.0
			for index in xrange(test_num):
				if len(train_index[index+900])>0:
					test_err=ssd_const_lr.ssd(rnn,testX[index],testY[index],False)
					test_err_total+=test_err
			test_err_total*=10
			test_results+='|%f'%(test_err_total)
			os.system('''echo '|%s' >> results_saved_%s_test_%d_%s_%.4f_%.4f_%.4f_%.4f.log'''%(test_err_total,param['dcrorcst'],param['numberOfSave'],param['mode'],param['learnRates']['U'],param['learnRates']['W'],param['learnRates']['V'],param['learnRates']['s']))

		err_total=0.0
		for index in xrange(train_num):
			if len(train_index[index])>0:
				err=ssd_const_lr.ssd(rnn,trainX[index],trainY[index])
				err_total+=err
			if index%param['batch']==0:
				rnn.update(param['learnRates'],lr_decay)
		rnn.update(param['learnRates'],lr_decay)
		results+='|%f'%err_total
		os.system('''echo '|%s' >> results_saved_%s_train_%d_%s_%.4f_%.4f_%.4f_%.4f.log'''%(err_total,param['dcrorcst'],param['numberOfSave'],param['mode'],param['learnRates']['U'],param['learnRates']['W'],param['learnRates']['V'],param['learnRates']['s']))
		print 'epoch %d completed!'%epoch
		if maketest:
			print 'total train error=%f'%err_total
			print 'total test error=%f'%test_err_total
		else:
			print 'total error=%f'%err_total
	results+='\n'
	end=time.time()
	results+='time=%.2f\n'%(end-begin)

	if maketest:
		test_results+='\n'
		test_results+='time=%.2f\n'%(end-begin)

######################
#                    #
#   Start RMSprop    #
#                    #
######################

if param['mode']=='sgd_rms':
	print 'Not implemented yet'
	
######################
#                    #
#   Start RMSspec    #
#                    #
######################

if param['mode']=='ssd_rms':
	begin=time.time()
	lr_decay=1
	for epoch in xrange(param['nEpoch']):
		if decreasingLR:
			lr_decay=epoch+1
		if maketest:
			test_err_total=0.0
			for index in xrange(test_num):
				if len(train_index[index+900])>0:
					test_err=ssd_rms.rms(rnn,testX[index],testY[index],param['alpha'],False)
					test_err_total+=test_err
			test_err_total*=10
			test_results+='|%f'%(test_err_total)
			os.system('''echo '|%s' >> results_saved_%s_test_%d_%s_%.4f_%.4f_%.4f_%.4f_%.4f.log'''%(test_err_total,param['dcrorcst'],param['numberOfSave'],param['mode'],param['learnRates']['U'],param['learnRates']['W'],param['learnRates']['V'],param['learnRates']['s'],param['alpha']))

		err_total=0.0
		for index in xrange(train_num):
			if len(train_index[index])>0:
				err=ssd_rms.rms(rnn,trainX[index],trainY[index],param['alpha'])
				err_total+=err
			if index%param['batch']==0:
				rnn.update(param['learnRates'],epochh+1)  
		rnn.update(param['learnRates'],epochh+1)
		results+='|%f'%err_total
		os.system('''echo '|%s' >> results_saved_%s_train_%d_%s_%.4f_%.4f_%.4f_%.4f_%.4f.log'''%(err_total,param['dcrorcst'],param['numberOfSave'],param['mode'],param['learnRates']['U'],param['learnRates']['W'],param['learnRates']['V'],param['learnRates']['s'],param['alpha']))
		print 'epoch %d completed!'%epoch
		if maketest:
			print 'total train error=%f'%err_total
			print 'total test error=%f'%test_err_total
		else:
			print 'total error=%f'%err_total
	results+='\n'
	end=time.time()
	results+='time=%.2f\n'%(end-begin)

	if maketest:
		test_results+='\n'
		test_results+='time=%.2f\n'%(end-begin)

######################
#                    #
#   Start ADAprop    #
#                    #
######################

if param['mode']=='sgd_adagrad':
	print 'Not implemented yet'

######################
#                    #
#   Start ADAspec    #
#                    #
######################
if param['mode']=='ssd_adagrad':
	print 'Not implemented yet'




with open(param['outfile'],param['outputMode']) as fopen:
	fopen.write(results)

if maketest:
	with open(param['test_outfile'],param['outputMode']) as fopen:
		fopen.write(test_results)

File: Torch/tasks/atisLabel.lua
require 'math'

require 'util.gradient'
require 'util.loader'
require 'util.matrixNorm'
require 'util.randSVD'
require 'util.sharp'
require 'util.vectorNorm'
require 'models.RNN'
require 'optimization.sgd'
require 'optimization.ssd'
require 'optimization.ssd_simp'

--hyper-parameters
param={}
param.trainXFile='../atis/train_word1000.csv'
param.trainYFile='../atis/train_label1000.csv'
param.testXFile='../atis/test_word50.csv'
param.testYFile='../atis/test_label50.csv'
param.dictFile='../atis/dict10.csv'
param.vectorDim=10
param.window=1
param.hiddens=10
param.outputs=128
param.batch=50
param.epoch=3
param.sgdLearningRateUWV=5
param.sgdUWVDecay=1
param.sgdLearningRateS=2
param.sgdSDecay=1
param.ssdLearningRateUWV=5
param.ssdUWVDecay=1
param.ssdLearningRateS=2
param.ssdSDecay=1
param.inputs=param.window*param.vectorDim
param.leftPad=math.floor(param.window/2)

trainIndex=loader:loadData(param.trainXFile)
trainLabel=loader:loadData(param.trainYFile)
testIndex=loader:loadData(param.testXFile)
testLabel=loader:loadData(param.testYFile)
dict=loader:loadDict(param.dictFile)
dict[-1]=torch.rand(param.vectorDim)*0.05		--padding vector

--create 2 RNNs that have the same parameters and share the intital states
rnn1={i2h=nil,h2h=nil,h2o=nil,s=nil,ds=nil,buffer=0,__init__=RNN.__init__,
run1Token=RNN.run1Token,runTokens=RNN.runTokens,update=RNN.update}
rnn2={i2h=nil,h2h=nil,h2o=nil,s=nil,ds=nil,buffer=0,__init__=RNN.__init__,
run1Token=RNN.run1Token,runTokens=RNN.runTokens,update=RNN.update}
rnn3={i2h=nil,h2h=nil,h2o=nil,s=nil,ds=nil,buffer=0,__init__=RNN.__init__,
run1Token=RNN.run1Token,runTokens=RNN.runTokens,update=RNN.update}
rnn1:__init__(param.inputs,param.hiddens,param.outputs)
rnn2:__init__(param.inputs,param.hiddens,param.outputs)
rnn3:__init__(param.inputs,param.hiddens,param.outputs)
rnn2.s:copy(rnn1.s)
rnn2.i2h.weight:copy(rnn1.i2h.weight)
rnn2.h2h.weight:copy(rnn1.h2h.weight)
rnn2.h2o.weight:copy(rnn1.h2o.weight)
rnn3.s:copy(rnn1.s)
rnn3.i2h.weight:copy(rnn1.i2h.weight)
rnn3.h2h.weight:copy(rnn1.h2h.weight)
rnn3.h2o.weight:copy(rnn1.h2o.weight)

--create the input and expected output of RNN
trainNum=table.getn(trainIndex)
testNum=table.getn(testIndex)
trainX={}
trainY={}
testX={}
testY={}
for i=1,trainNum do
	if i%100==0 then
		print(i,'/',trainNum)
	end
	local sentenceLen=table.getn(trainIndex[i])
	local singleInput=torch.zeros(sentenceLen,param.inputs)
	local singleLabel=torch.zeros(sentenceLen,param.outputs)
	for p=2,sentenceLen+param.window do
		local toFill=dict[-1]
		if p>param.leftPad+1 and p<=sentenceLen+param.leftPad+1 then
			local word=trainIndex[i][p-param.leftPad-1]
			if dict[word]==nil then
				dict[word]=torch.rand(param.vectorDim)*0.05
			end
			toFill=dict[word]
		end
		local startline=math.max(1,p-param.window)
		local endline=math.min(sentenceLen,p-1)
		for line=startline,endline do
			local col=p-line
			singleInput[{{line,line},{(col-1)*param.vectorDim+1,col*param.vectorDim}}]=toFill:reshape(1,param.vectorDim)
		end
	end
	for w=1,sentenceLen do
		singleLabel[w][trainLabel[i][w]+1]=1
	end
	table.insert(trainX,singleInput)
	table.insert(trainY,singleLabel)
end

for i=1,testNum do
	if i%100==0 then
		print(i,'/',testNum)
	end
	local sentenceLen=table.getn(testIndex[i])
	local singleInput=torch.zeros(sentenceLen,param.inputs)
	local singleLabel=torch.zeros(sentenceLen,param.outputs)
	for p=2,sentenceLen+param.window do
		local toFill=dict[-1]
		if p>param.leftPad+1 and p<=sentenceLen+param.leftPad+1 then
			local word=testIndex[i][p-param.leftPad-1]
			if dict[word]==nil then
				dict[word]=torch.rand(param.vectorDim)*0.05
			end
			toFill=dict[word]
		end
		local startline=math.max(1,p-param.window)
		local endline=math.min(sentenceLen,p-1)
		for line=startline,endline do
			local col=p-line
			singleInput[{{line,line},{(col-1)*param.vectorDim+1,col*param.vectorDim}}]=toFill:reshape(1,param.vectorDim)
		end
	end
	for w=1,sentenceLen do
		singleLabel[w][testLabel[i][w]+1]=1
	end
	table.insert(testX,singleInput)
	table.insert(testY,singleLabel)
end

--SGD--
local sgd_lr_uwv=param.sgdLearningRateUWV
local sgd_lr_s=param.sgdLearningRateS
for iter=1,param.epoch do
	local err_total=0
	for i=1,trainNum do
		local len=table.getn(trainIndex[i])
		if len>0 then
			local err=sgd(rnn1,trainX[i],trainY[i])
			err_total=err_total+err
		end
		if i%param.batch==0 then
			-- print(vectorNorm:norm(rnn1.i2h.gradWeight,2)/rnn1.i2h.gradWeight:nElement())
			-- print(vectorNorm:norm(rnn1.h2h.gradWeight,2)/rnn1.h2h.gradWeight:nElement())
			-- print(vectorNorm:norm(rnn1.h2o.gradWeight,2)/rnn1.h2o.gradWeight:nElement())
			-- print(vectorNorm:norm(rnn1.ds,2)/rnn1.ds:nElement())
			-- print(vectorNorm:norm(rnn1.s,2)/rnn1.s:nElement())
			-- print('------------------------------')
			-- print(vectorNorm:norm(rnn1.h2o.weight,2)/rnn1.h2o.weight:nElement())
			rnn1:update(sgd_lr_uwv,sgd_lr_s)
			print('epoch=%d,index=%d/%d'%{iter,i,trainNum})
			print('accumulated error=%f'%err_total)
			local u1,s1,v1=torch.svd(rnn1.i2h.weight)
			local u2,s2,v2=torch.svd(rnn1.h2h.weight)
			local u3,s3,v3=torch.svd(rnn1.h2o.weight)
			-- print(s1:reshape(1,s1:size(1)))
			-- print(s2:reshape(1,s2:size(1)))
			-- print(s3:reshape(1,s3:size(1)))
		end
	end
	rnn1:update(sgd_lr_uwv,sgd_lr_s)
	print('epoch %d completed'%{iter})
	print('total error=%f'%err_total)
	sgd_lr_uwv=sgd_lr_uwv*param.sgdUWVDecay
	sgd_lr_s=sgd_lr_s*param.sgdSDecay
end

--SSD--
local ssd_lr_uwv=param.ssdLearningRateUWV
-- local ssd_lr_s=param.ssdLearningRateS
-- for iter=1,param.epoch do
-- 	local err_total=0
-- 	for i=1,trainNum do
-- 		local len=table.getn(trainIndex[i])
-- 		if len>0 then
-- 			local err=ssd(rnn2,trainX[i],trainY[i])
-- 			err_total=err_total+err
-- 		end
-- 		-- print(vectorNorm:norm(rnn2.i2h.gradWeight,2))
-- 		-- print(vectorNorm:norm(rnn2.h2h.gradWeight,2))
-- 		-- print(vectorNorm:norm(rnn2.h2o.gradWeight,2))
-- 		-- print(vectorNorm:norm(rnn2.ds,2))
-- 		if i%param.batch==0 then
-- 			rnn2:update(ssd_lr_uwv,ssd_lr_s)
-- 			print('epoch=%d,index=%d/%d'%{iter,i,trainNum})
-- 			print('accumulated error=%f'%err_total)
-- 		end
-- 	end
-- 	rnn2:update(ssd_lr_uwv,ssd_lr_s)
-- 	print('epoch %d completed'%iter)
-- 	print('total error=%f'%err_total)
-- 	ssd_lr_uwv=ssd_lr_uwv*param.ssdUWVDecay
-- 	ssd_lr_s=ssd_lr_s*param.ssdSDecay
-- end

--SSD_Simp
local ssd_simp_lr_uwv=5
local ssd_simp_lr_s=2
for iter=1,param.epoch do
	local err_total=0
	for i=1,trainNum do
		local len=table.getn(trainIndex[i])
		if len>0 then
			local err=ssd_simp(rnn3,trainX[i],trainY[i])
			err_total=err_total+err
		end
		-- print(vectorNorm:norm(rnn2.i2h.gradWeight,2))
		-- print(vectorNorm:norm(rnn2.h2h.gradWeight,2))
		-- print(vectorNorm:norm(rnn2.h2o.gradWeight,2))
		-- print(vectorNorm:norm(rnn2.ds,2))
		if i%param.batch==0 then
			rnn2:update(ssd_simp_lr_uwv,ssd_simp_lr_s)
			print('epoch=%d,index=%d/%d'%{iter,i,trainNum})
			print('accumulated error=%f'%err_total)
		end
	end
	rnn2:update(ssd_simp_lr_uwv,ssd_simp_lr_s)
	print('epoch %d completed'%iter)
	print('total error=%f'%err_total)
	ssd_simp_lr_uwv=ssd_simp_lr_uwv*1
	ssd_simp_lr_s=ssd_simp_lr_s*1
end
